query-service-impl/src/main/java/org/hypertrace/core/query/service/projection/ProjectionTransformation.java

```diff
@@ -0,0 +1,282 @@
+import javax.inject.Inject;
+import org.hypertrace.core.attribute.service.cachingclient.CachingAttributeClient;
+import org.hypertrace.core.attribute.service.v1.AttributeDefinition;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadata;
+import org.hypertrace.core.attribute.service.v1.LiteralValue;
+import org.hypertrace.core.attribute.service.v1.Projection;
+import org.hypertrace.core.attribute.service.v1.ProjectionExpression;
+import org.hypertrace.core.attribute.service.v1.ProjectionOperator;
+import org.hypertrace.core.query.service.QueryFunctionConstants;
+import org.hypertrace.core.query.service.QueryTransformation;
+import org.hypertrace.core.query.service.api.ColumnIdentifier;
+import org.hypertrace.core.query.service.api.Expression;
+import org.hypertrace.core.query.service.api.Filter;
+import org.hypertrace.core.query.service.api.Function;
+import org.hypertrace.core.query.service.api.OrderByExpression;
+import org.hypertrace.core.query.service.api.QueryRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class ProjectionTransformation implements QueryTransformation {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProjectionTransformation.class);
+
+  private final CachingAttributeClient attributeClient;
+
+  @Inject
+  ProjectionTransformation(CachingAttributeClient attributeClient) {
+    this.attributeClient = attributeClient;
+  }
+
+  @Override
+  public Single<QueryRequest> transform(
+      QueryRequest queryRequest, QueryTransformationContext transformationContext) {
+    return zip(
+            this.transformExpressionList(queryRequest.getSelectionList()),
+            this.transformExpressionList(queryRequest.getAggregationList()),
+            this.transformFilter(queryRequest.getFilter()),
+            this.transformExpressionList(queryRequest.getGroupByList()),
+            this.transformOrderByList(queryRequest.getOrderByList()),
+            (selections, aggregations, filter, groupBys, orderBys) ->
+                this.rebuildRequestOmittingDefaults(
+                    queryRequest, selections, aggregations, filter, groupBys, orderBys))
+        .doOnSuccess(transformed -> this.debugLogIfRequestTransformed(queryRequest, transformed));
+  }
+
+  private Single<List<Expression>> transformExpressionList(List<Expression> expressionList) {
+    return Observable.fromIterable(expressionList)
+        .flatMapSingle(this::transformExpression)
+        .toList();
+  }
+
+  private Single<Expression> transformExpression(Expression expression) {
+    switch (expression.getValueCase()) {
+      case COLUMNIDENTIFIER:
+        return this.transformColumnIdentifier(expression.getColumnIdentifier());
+      case FUNCTION:
+        return this.transformFunction(expression.getFunction())
+            .map(expression.toBuilder()::setFunction)
+            .map(Expression.Builder::build);
+      case ORDERBY:
+        return this.transformOrderBy(expression.getOrderBy())
```

Comment:
74-76 not unit tested

Reply:
So I actually meant to ask about this. I'm unclear if this is a missing test, or if this an issue in the modeling? In what scenario is an expression of type orderBy?  Order by is set at the top level of the query as an expression, which can be a function or column identifier. The request from the order by test looks like:

```java
    QueryRequest originalRequest =
        QueryRequest.newBuilder()
            .addSelection(createColumnExpression(PROJECTED_ATTRIBUTE_ID))
            .addOrderBy(createOrderByExpression(PROJECTED_ATTRIBUTE_ID, SortOrder.DESC))
            .build();
```
where the order by expression is built by:
```java
  public static OrderByExpression.Builder createOrderByExpression(
      String columnName, SortOrder order) {
    return OrderByExpression.newBuilder()
        .setOrder(order)
        .setExpression(createColumnExpression(columnName));
  }
```

Reply:
I see what you are saying. Seems like a problem with the API def.